rcl_logging_implementation/src/functions.cpp

```diff
@@ -0,0 +1,270 @@
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "failed to resolve symbol '%s' in shared library due to %s",
+        symbol_name.c_str(), e.what());
+    }
+    return nullptr;
+  }
+  return lib->get_symbol(symbol_name);
+}
+
+void *
+get_logging_symbol(const char * symbol_name)
+{
+  try {
+    return lookup_logging_symbol(get_logging_library(), symbol_name);
+  } catch (const std::exception & e) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "failed to get symbol '%s' due to %s",
+      symbol_name, e.what());
+    return nullptr;
+  }
+}
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define CALL_SYMBOL(symbol_name, ReturnType, error_value, ...) \
+  if (!symbol_ ## symbol_name) { \
+    symbol_ ## symbol_name = get_logging_symbol(#symbol_name); \
+  } \
+  if (!symbol_ ## symbol_name) { \
+    /* error message set by get_logging_symbol() */ \
+    return error_value; \
+  } \
+  typedef ReturnType (* FunctionSignature)(__VA_ARGS__); \
+  FunctionSignature func = reinterpret_cast<FunctionSignature>(symbol_ ## symbol_name); \
+  return func
+
+// Symbol pointers for lazy loading
+void * symbol_rcl_logging_external_initialize = nullptr;
+void * symbol_rcl_logging_external_shutdown = nullptr;
+void * symbol_rcl_logging_external_log = nullptr;
+void * symbol_rcl_logging_external_set_logger_level = nullptr;
+
+rcl_logging_ret_t
+rcl_logging_external_initialize(
+  const char * file_name_prefix,
+  const char * config_file,
+  rcutils_allocator_t allocator)
+{
+  RCUTILS_LOG_DEBUG_NAMED(
+    "rcl_logging_implementation",
+    "rcl_logging_external_initialize called (prefix: %s, config: %s)",
+    file_name_prefix ? file_name_prefix : "NULL", config_file ? config_file : "NULL");
+  CALL_SYMBOL(
+    rcl_logging_external_initialize,
+    rcl_logging_ret_t,
+    RCL_LOGGING_RET_ERROR,
+    const char *, const char *, rcutils_allocator_t)(file_name_prefix, config_file, allocator);
+}
+
+rcl_logging_ret_t
+rcl_logging_external_shutdown(void)
+{
+  RCUTILS_LOG_DEBUG_NAMED(
+    "rcl_logging_implementation",
+    "rcl_logging_external_shutdown called");
+  if (!symbol_rcl_logging_external_shutdown) {
+    symbol_rcl_logging_external_shutdown = get_logging_symbol("rcl_logging_external_shutdown");
+  }
+  if (!symbol_rcl_logging_external_shutdown) {
+    // If shutdown is called before init, it's OK to just return success
+    return RCL_LOGGING_RET_OK;
+  }
+
+  typedef rcl_logging_ret_t (* FunctionSignature)(void);
+  FunctionSignature func = reinterpret_cast<FunctionSignature>(
+    symbol_rcl_logging_external_shutdown);
+  rcl_logging_ret_t ret = func();
+
+  // Unload the library after successful shutdown
+  unload_logging_library();
+
+  return ret;
+}
+
+void
+rcl_logging_external_log(int severity, const char * name, const char * msg)
+{
+  if (!symbol_rcl_logging_external_log) {
+    symbol_rcl_logging_external_log = get_logging_symbol("rcl_logging_external_log");
+  }
+  if (!symbol_rcl_logging_external_log) {
```

Comment:
The rcl_logging_external_log() function is called very frequently, and it performs two checks each time. Is there any way to optimize this?  

One of my ideas is  
For `rcl_logging_external_shutdown`, `rcl_logging_external_log`, and `logging_external_set_logger_level`, use function pointers. Initially, all of them point to default functions, which can simply output uninitialized error.

When the user calls `rcl_logging_external_initialize()`, except calling the rcl_logging_external_initialize() in the logging library, the three function pointers are also updated to point to the corresponding functions in the logging library. This way, when the user calls these three functions, they directly invoke the function pointers.   

Do you think this approach is feasible?